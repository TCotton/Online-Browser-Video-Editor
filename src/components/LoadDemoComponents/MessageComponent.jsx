import React, {useCallback} from 'react';
import {window} from "browser-monads";
import {useDispatch} from "react-redux";
import happyPanda from '../../../static/happy-panda.mp4'
import db from "../indexDB/indexDB";
import dbVF from "../indexDB/indexDBVideo";
import {displayFn} from '../slices/loaderSlice'

export const dex = {
    dexieRun: (files) => {
        db.file.clear();
        db.file.add(files);
    },
    dexieRunVideo: (file) => {
        dbVF.videofile.clear();
        dbVF.videofile.add(file);
    }
}

export const MessageComponent = () => {

    const dispatch = useDispatch();

    const onClickFnc = useCallback((e) => {

        dispatch(displayFn(true));

        window.fetch(happyPanda)
            .then(response => {
                if (response.status === 200) {
                    return response.blob();
                }
            }).then((blob) => {

            const file = {
                lastModified: Date.now(),
                name: "happypanda.mp4",
                size: blob.size,
                type: blob.type
            }
            dex.dexieRunVideo(file);
            dex.dexieRun(blob);

        }).catch((error) => {
            throw new Error(error.toString());
        });

        e.preventDefault();
    }, []);

    return (
        <section className="message" data-testid="message">
            <p>
                Upload and transcode video files for use on social media.
            </p>
            <p>This is a lockdown project by <a href="https://andywalpole.me/blog/" target="_blank"
                                                rel="noopener noreferrer">Andrew Walpole</a>, built
                using D3.js, the Web Audio API, IndexedDB and WebAssembly (WASM) on a Gatsby/React/Redux base. <a
                    href="/" className="demo" onClick={onClickFnc} data-testid='link'>Click here to load a demo video
                    file</a>. <a
                    href="https://github.com/TCotton/Online-Browser-Video-Editor" target="_blank"
                    rel="noopener noreferrer">This is an alpha release</a>.
            </p>
        </section>
    )
}