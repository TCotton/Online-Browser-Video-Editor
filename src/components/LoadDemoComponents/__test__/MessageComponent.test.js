import React from 'react';
import fetchMock from 'fetch-mock';
import {render, waitFor, fireEvent} from "@testing-library/react";
import { enableFetchMocks } from 'jest-fetch-mock'
import {MessageComponent, dex} from "../MessageComponent";

enableFetchMocks();

describe('MessageComponent', () => {

    const mockDexieRun = jest.fn(n => n);
    const mockDexieRunVideo = jest.fn(n => n);

    beforeEach(() => {
       dex.dexieRun = mockDexieRun;
       dex.dexieRunVideo = mockDexieRunVideo;
    });

  /*  afterEach(() => {
        mockDexieRun.reset();
        mockDexieRunVideo.reset();
    })*/

    it('should render successfully', () => {
        const {container} = render(<MessageComponent />);
        expect(container.querySelector("[data-testid=\"message\"]")).not.toBeNull();
    });

    it('should match snapshot', () => {
        const {baseElement} = render(<MessageComponent />);
        expect(baseElement).toMatchSnapshot();
    });

    it('should call dexieRun and dexieRunVideo on clicking link',  async () => {
        const blob = new Blob(['a', 'b', 'c', 'd']);
        fetchMock.once('*', {body: blob}, {sendAsJson: false});
        const {container} = render(<MessageComponent />);
        const link = container.querySelector("[data-testid=\"link\"]")

        await waitFor(() => {
            fireEvent.click(link);
        })

        expect(mockDexieRun).toHaveBeenCalledTimes(1);
        expect(mockDexieRunVideo).toHaveBeenCalledTimes(1)
    });
});