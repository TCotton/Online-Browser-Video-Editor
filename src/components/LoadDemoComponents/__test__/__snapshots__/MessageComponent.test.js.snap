// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageComponent should match snapshot 1`] = `
<body>
  <div>
    <section
      class="message"
      data-testid="message"
    >
      <p>
        Upload and transcode video files for use on social media.
      </p>
      <p>
        This is a lockdown project by
         
        <a
          href="https://andywalpole.me/blog/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Andrew Walpole
        </a>
        , built using D3.js, the Web Audio API, IndexedDB and WebAssembly (WASM) on a Gatsby/React/Redux base.
         
        <a
          class="demo"
          data-testid="link"
          href="/"
        >
          Click here to load a demo video file
        </a>
        .
         
        <a
          href="https://github.com/TCotton/Online-Browser-Video-Editor"
          rel="noopener noreferrer"
          target="_blank"
        >
          This is an alpha release
        </a>
        .
      </p>
    </section>
  </div>
</body>
`;
